### PR TITLE
Add SoC blog post

### PR DESCRIPTION
**Pull request by lucifer4j**
Final Work Product for submission in Google Summer of Code 2023 - Eclipse Vert.x client for Apache Pinot. Please do merge this PR yet. There are multiple steps related to transferring the repository and publishing a release left before the blog post is published.

cc: @tsegismont

### CODE REVIEW

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 1]
```diff
@@ -0,0 +1,120 @@
+---
+title: Vert.x client for Apache Pinot
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
title: Eclipse Vert.x client for Apache Pinot
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+category: News
+authors:
+  - name: Lucifer Morningstar
+    github_id: lucifer4j
+summary: >-
+    Apache Pinot is a realtime distributed datastore for analytics workloads. To make
+    it easier for vert.x users to use Apache Pinot in their applications, we introduce
+    a vertx-pinot-client extension.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
    Apache Pinot is a realtime distributed datastore for analytics workloads. To make
    it easier for Eclipse Vert.x users to use Apache Pinot in their applications, we introduce
    the Reactiverse Vert.x Client for Pinot.
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+---
+
+<Alert warning>
+
+The blog post is informational and a teaser of upcoming changes to *vertx-pinot-client* which is under active
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
The blog post is informational and a teaser of upcoming changes to `pinot-client` which is under active
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+</Alert>
+
+## Introduction
+
+Eclipse Vert.x is a toolkit to build reactive applications on the Java virtual machine. It provides asynchronous and
+non-blocking clients for different types of databases. Apache Pinot is a realtime distributed datastore for analytics
+workloads. The [Apache Pinot Java Client](https://docs.pinot.apache.org/users/clients/java) uses
+[AsyncHttpClient](https://github.com/AsyncHttpClient/async-http-client) as the default transport and is cumbersome to
+use in Eclipse Vert.x applications.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]: Can you rephrase this? For example, you could say "unpractical" instead of "cumbersome" and say why (blocking API implies to wrap everything with `executeBlocking`.

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+This project proposes to create a vertx-pinot-client extension. This extension will be a lightweight wrapper around
+the existing Pinot Java client. It will also offer alternative transport for the client using vertx-web-client.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
The [`pinot-client`](https://github.com/reactiverse/pinot-client) Reactiverse project is a lightweight wrapper around the existing Pinot Java client. It also provides an alternative transport using `vertx-web-client`.
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+This will make the Apache Pinot more accessible to Vert.x users.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx:29]: Here you can explain the advantages:

- async and non-blocking API
- RxJava and Mutiny bindings thanks to Vert.x codegen
- better performances by using the the Vert.x Web Client (thus allowing to stay on the same event loop when interacting with Pinot)

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+## Usage
+
+Add the pinot-client dependency to your project.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
Add the `pinot-client` dependency to your project.
```

Using back ticks allows the reader to visualize code related text immediately.

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+You can create an instance of the client to interact with the Pinot cluster as follows.
+
+```java
+String brokerUrl = "localhost:8000";
+VertxPinotClientTransport transport = new VertxPinotClientTransport(vertx);
+VertxConnection connection = VertxConnectionFactory.fromHostList(vertx, List.of(brokerUrl), transport);
+```
+
+Now we can use the Connection to query the cluster. Here is an example to query the top 10 home run scorers from
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
Now we can use the `Connection` to query the cluster. Here is an example to query the top 10 home run scorers from
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+to monitors their orders in realtime. The dashboard deals with three types of data: products, users, and orders. The
+dashboard uses the Vert.x Pinot Client to [`query`](https://github.com/lucifer4j/pizza-shop-demo/blob/main/dashboard/backend/src/main/java/pizzashop/DashboardVerticle.java)
+the Pinot cluster.
+
+![Order Dashboard Statistics](/images/blog/soc-vertx-pinot-client/demo-dashboard-1.png)
+![Order Dashboard Latest and Popular Items](/images/blog/soc-vertx-pinot-client/demo-dashboard-2.png)
+
+## Challenges
+
+While working on the project, I encountered several implementation challenges.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]: You could expand this section explaining that you created this project as part of the GSoC program

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+#### package-private `org.apache.pinot.client.BrokerResponse` class
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
#### Package-private `org.apache.pinot.client.BrokerResponse` class
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+methods in the `PinotClientTransport` return this type. This makes it difficult to implement the interface in an external package.
+
+To work around this issue, we created a `org.apache.pinot.client` package inside our wrapper project. The [`VertxPinotClientTransport`](https://github.com/lucifer4j/pinot-client/blob/70dc4152cc77fd6ae521f94fb8c59bf4c7f9d12a/pinot-client/src/main/java/org/apache/pinot/client/VertxPinotClientTransport.java#L37)
+implementation currently lives inside this package for the same reason. This workaround only works for projects running in `classpath`
+mode but not for projects using `modulepath` because Java Modules disallow package splitting. To fix the issue as source,
+I opened a [`PR`](https://github.com/apache/pinot/pull/10932) in `apache/pinot` repository. The PR is now merged. Once a new
+release of Pinot is available, I will be updating the client wrapper as well.
+
+#### missing `ConnectionFactory` method
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
#### Missing `ConnectionFactory` method
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+constructors of the `Connection` class are package-private preventing us from directly adding the missing method in our
+wrapper.
+
+Again to fix this issue, I created a [`bridge class`](https://github.com/lucifer4j/pinot-client/blob/70dc4152cc77fd6ae521f94fb8c59bf4c7f9d12a/pinot-client/src/main/java/org/apache/pinot/client/ConnectionFactoryBridge.java#L28)
+in the `org.apache.pinot.client` package. Simultaneously, I opened a [`PR`](https://github.com/apache/pinot/pull/11013) in
+`apache/pinot` repository. The PR is also now merged and the Vert.x client will be updated once a new release is available.
+
+## Bonus Fixes
+
+#### missing double-checked locking in ConnectionFactory
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
#### Missing double-checked locking in `ConnectionFactory`
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+While working on the tasks outlined earlier, I (more accurately my IDE's static checking) discovered the double-checked
+locking mechanism in `ConnectionFactory` class in `pinot-client` was broken. I opened a [`PR`](https://github.com/apache/pinot/pull/11014) to fix it.
+
+#### broken docker health checks in pizza-shop-demo
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
#### Broken Docker health checks in `pizza-shop-demo`
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+The Docker Health Checks in the original pizza-shop-demo were broken. The health checks had been written using `nc` utility
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
The Docker Health Checks in the original `pizza-shop-demo` were broken. The health checks had been written using `nc` utility
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+A more real-life demo is available at [`pizza-shop-demo`](https://github.com/lucifer4j/pizza-shop-demo). It includes a
+pizza delivery service specializing in pizzas with Indian toppings. They have a dashboard built using Apache Pinot
+to monitors their orders in realtime. The dashboard deals with three types of data: products, users, and orders. The
+dashboard uses the Vert.x Pinot Client to [`query`](https://github.com/lucifer4j/pizza-shop-demo/blob/main/dashboard/backend/src/main/java/pizzashop/DashboardVerticle.java)
+the Pinot cluster.
+
+![Order Dashboard Statistics](/images/blog/soc-vertx-pinot-client/demo-dashboard-1.png)
+![Order Dashboard Latest and Popular Items](/images/blog/soc-vertx-pinot-client/demo-dashboard-2.png)
+
+## Challenges
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx:86]: Perhaps you could also mention the problem with blocking java `Future`? Also, how you solved it temporarily with `executeBlocking`, and the conversation with Pinot commiters to get the pending PR for `CompletableFuture` merged. And how it will remove the need for offloading to the Vert.x worker pool.

**Review by tsegismont — COMMENTED** [review]
Thank you @lucifer4j , here's a batch of comments

**Comment by tsegismont** [comment]
Thanks for applying suggestions. Any idea about when you'll be able to address the last comments?

File: blog/2023-08-21-soc-vertx-pinot-client.mdx (hunk 6 above)
Comment (lucifer4j) [reply blog/2023-08-21-soc-vertx-pinot-client.mdx:29]: done

File: blog/2023-08-21-soc-vertx-pinot-client.mdx (hunk 15 above)
Comment (lucifer4j) [reply blog/2023-08-21-soc-vertx-pinot-client.mdx:86]: done

File: blog/2023-08-21-soc-vertx-pinot-client.mdx (hunk 4 above)
Comment (lucifer4j) [reply blog/2023-08-21-soc-vertx-pinot-client.mdx]: done

File: blog/2023-08-21-soc-vertx-pinot-client.mdx (hunk 9 above)
Comment (lucifer4j) [reply blog/2023-08-21-soc-vertx-pinot-client.mdx]: done

**Comment by lucifer4j** [comment]
Hi! I have done the required changes.

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+methods in the `PinotClientTransport` return this type. This makes it difficult to implement the interface in an external package.
+
+To work around this issue, we created a `org.apache.pinot.client` package inside our wrapper project. The [`VertxPinotClientTransport`](https://github.com/lucifer4j/pinot-client/blob/70dc4152cc77fd6ae521f94fb8c59bf4c7f9d12a/pinot-client/src/main/java/org/apache/pinot/client/VertxPinotClientTransport.java#L37)
+implementation currently lives inside this package for the same reason. This workaround only works for projects running in `classpath`
+mode but not for projects using `modulepath` because Java Modules disallow package splitting. To fix the issue as source,
+I opened a [`PR`](https://github.com/apache/pinot/pull/10932) in `apache/pinot` repository. The PR is now merged. Once a new
+release of Pinot is available, I will be updating the client wrapper as well.
+
+#### `java.concurrent.util.Future` return type
+Some methods of the `PinotClientTransport` [interface](https://github.com/apache/pinot/blob/6e235a4ec2a16006337da04e118a435b5bb8f6d8/pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/PinotClientTransport.java#L32)
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
#### `java.concurrent.util.Future` return type

Some methods of the `PinotClientTransport` [interface](https://github.com/apache/pinot/blob/6e235a4ec2a16006337da04e118a435b5bb8f6d8/pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/PinotClientTransport.java#L32)
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+return a `java.concurrent.util.Future` future. However, `Future.get` is a blocking call which is undesirable. To work around the
+issue, we wrap every request and query call in an [`executeBlocking` block](https://github.com/reactiverse/pinot-client/blob/main/pinot-client/src/main/java/io/reactiverse/pinot/client/impl/Utils.java#L24-L34).
+To avoid blocking calls together, we discussed replacing `Future` with `CompletabeFutre`. The same change has already been [merged](https://github.com/apache/pinot/pull/10326)
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
To avoid blocking calls altogether, we discussed replacing `Future` with `CompletabeFutre`. The same change has already been [merged](https://github.com/apache/pinot/pull/10326)
```

File: blog/2023-08-21-soc-vertx-pinot-client.mdx [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+now and will be available in the next release. Once that is available, we can update the wrapper client to adapt `CompletableFuture`s
+to `Vert.x Futures` without much hassle and without offloading the task to blocking worker pool.
```
Comment (tsegismont) [inline blog/2023-08-21-soc-vertx-pinot-client.mdx]:
```suggestion
to `Vert.x Futures` without much hassle and without having to offload some of the tasks to the Vert.x worker pool.
```

**Review by tsegismont — COMMENTED** [review]
Another round of suggestions, then I think we're good

**Comment by lucifer4j** [comment]
Hi! I have updated the PR with your changes and also added a short disclaimer on current release status.

**Review by tsegismont — APPROVED** [review]
Thank you @lucifer4j !

**Comment by tsegismont** [comment]
The blog post is online! https://vertx.io/blog/soc-vertx-pinot-client/
